manta-util/src/time/lock.rs

```diff
@@ -0,0 +1,205 @@
+    #[inline]
+    pub fn set(&mut self, value: T) -> T {
+        self.mutate(move |t| mem::replace(t, value))
+    }
+
+    /// Mutates the internal value using `f`, resetting the modification time to [`Instant::now`].
+    #[inline]
+    pub fn mutate<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut T) -> R,
+    {
+        self.mutate_timed(move |value, _| f(value))
+    }
+
+    /// Mutates the internal value with the [`Instant`] of the last modification to `self`,
+    /// resetting the modification time to [`Instant::now`].
+    #[inline]
+    pub fn mutate_timed<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut T, Instant) -> R,
+    {
+        let result = f(&mut self.value, self.instant);
+        self.tap();
+        result
+    }
+
+    /// Mutates the internal value using `f` if the amount of time since the last modification to
+    /// `self` was longer than `timeout`.
+    #[inline]
+    pub fn mutate_if_expired<F, R>(&mut self, timeout: Duration, f: F) -> Option<R>
+    where
+        F: FnOnce(&mut T) -> R,
+    {
+        if self.has_expired(timeout) {
+            Some(self.mutate(f))
+        } else {
+            None
+        }
+    }
+
+    /// Sets the internal value to `value` if the amount of time since the last modification to
+    /// `self` was longer than `timeout`.
+    #[inline]
+    pub fn set_if_expired(&mut self, timeout: Duration, value: T) -> Option<T> {
+        self.set_with_if_expired(timeout, move || value)
+    }
+
+    /// Sets the internal value using `value` if the amount of time since the last modification to
+    /// `self` was longer than `timeout`.
+    #[inline]
+    pub fn set_with_if_expired<F>(&mut self, timeout: Duration, value: F) -> Option<T>
+    where
+        F: FnOnce() -> T,
+    {
+        self.mutate_if_expired(timeout, move |t| mem::replace(t, value()))
+    }
+
+    /// Returns the underlying timed value, dropping `self`.
+    #[inline]
+    pub fn into_inner(self) -> T {
+        self.value
+    }
+
+    /// Returns the underlying timed value and its last modification time, dropping `self`.
+    #[inline]
+    pub fn into_pair(self) -> (T, Instant) {
+        (self.value, self.instant)
+    }
+}
+
+impl<T> AsRef<T> for Timed<T> {
+    #[inline]
+    fn as_ref(&self) -> &T {
+        self.get()
+    }
+}
+
+impl<T> Clone for Timed<T>
+where
+    T: Clone,
+{
+    /// Clones the underlying data, creating a new [`Timed`] object with a new creation time set to
+    /// the return value of [`Instant::now`].
+    #[inline]
+    fn clone(&self) -> Self {
+        Self::new(self.value.clone())
+    }
+}
+
+impl<T> Default for Timed<T>
+where
+    T: Default,
+{
+    /// Builds a new [`Timed`] object from the default value of `T` and the current time returned by
+    /// [`Instant::now`].
+    #[inline]
+    fn default() -> Self {
+        Self::new(Default::default())
+    }
+}
+
+impl<T> Deref for Timed<T> {
```

Comment:
Why not having `DerefMut`? Doing so, by calling `*object = something`, we set the value to be `something` and let the last modified time be `Instant::now()`. Is this because someone might get a mutable reference by calling `deref_mut` and set the value several seconds later but the last modified time is still the instant that `deref_mut` is called?

Reply:
Yeah, we want the time to be related to the time of the change, not the time of acquiring the mutable reference. There might be a way to do `DerefMut` but it's out of the scope for this PR. Not using `DerefMut` is definitely safer.